### PR TITLE
[rllib] Added vanilla policy gradient

### PR DESCRIPTION
**Pull request by alvkao58**
Implemented vanilla policy gradient in RLLib.

### CODE REVIEW

**Comment by richardliaw** [comment]
ok to test

File: python/ray/rllib/pg/pg.py
Comment (ericl) [inline python/ray/rllib/pg/pg.py]: Shall we drop the gae / filter / clipping stuff to simplify the example?

Comment (richardliaw) [reply python/ray/rllib/pg/pg.py]: yes I think that's a really good idea; @alvkao58 what are your thoughts?

Comment (alvkao58) [reply python/ray/rllib/pg/pg.py]: Sure, that sounds good, I think I'll also drop the stuff relevant to the value function for sake of simplicity.

**Comment by ericl** [comment]
Could you add a test entry in https://github.com/ray-project/ray/blob/master/python/ray/rllib/test/test_supported_spaces.py ?

File: python/ray/rllib/pg/pg.py
Comment (ericl) [inline python/ray/rllib/pg/pg.py]: """Simple policy gradient agent.

This is an example agent to show how to implement algorithms in RLlib. In most cases, you will probably want
to use the PPO agent instead."""

Comment (ericl) [reply python/ray/rllib/pg/pg.py]: Could you also switch the link in http://ray.readthedocs.io/en/latest/rllib-dev.html from A3CEvaluator to PGAgent instead? (rllib-dev.rst file under doc/source)

Comment (alvkao58) [reply python/ray/rllib/pg/pg.py]: Done; I switched the link to be to PGEvaluator instead of PGAgent though (since it's supposed to be an example of an evaluator class).

File: python/ray/rllib/pg/pg.py
Comment (ericl) [inline python/ray/rllib/pg/pg.py]: This should go inside the class def (class doc)

File: doc/source/rllib-dev.rst
Comment (ericl) [inline doc/source/rllib-dev.rst]: see the simple policy gradient evaluator example

File: python/ray/rllib/pg/pg_evaluator.py
Comment (ericl) [inline python/ray/rllib/pg/pg_evaluator.py]: Evaluator for

File: python/ray/rllib/pg/pg.py
Comment (richardliaw) [inline python/ray/rllib/pg/pg.py]: hm I have to double check this but perhaps this `if-else` doesn't actually need to be here

File: python/ray/rllib/pg/policy.py
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: consider `reduce_mean`

File: python/ray/rllib/pg/policy.py
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: we don't need a bracket around `self.sample` right? this would simplify the indenting below.

File: python/ray/rllib/pg/policy.py
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: consider `feed_dict = dict(zip(self.grads, grads))`

File: python/ray/rllib/pg/policy.py
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: this isn't really used right? i would be ok with dropping it

File: python/ray/rllib/pg/policy.py
Comment (richardliaw) [inline python/ray/rllib/pg/policy.py]: I'm not so sure about the use of `worker_device` here - I can take a look into this, and chances are we don't actually need it.

File: python/ray/rllib/pg/pg.py
Comment (richardliaw) [inline python/ray/rllib/pg/pg.py]: filters are dropped right?

File: python/ray/rllib/pg/pg_evaluator.py
Comment (richardliaw) [inline python/ray/rllib/pg/pg_evaluator.py]: we don't need this right?

**Review by richardliaw — COMMENTED** [review]
Overall this is solid! Left a few comments. I think we should have a `tuned_example` (like on CartPole` or something.

File: python/ray/rllib/pg/policy.py
Comment (ericl) [reply python/ray/rllib/pg/policy.py]: Yeah you don't want to do this, TF will pick devices better by itself.

Comment (ericl) [reply python/ray/rllib/pg/policy.py]: You also shouldn't need the tf.Graph thing, agents already set up a default graph. In fact this might mess up checkpointing.

**Review by ericl — COMMENTED** [review]
Probably should also add this to the test checkpoint restore test in Rllib/test

File: python/ray/rllib/pg/policy.py
Comment (richardliaw) [reply python/ray/rllib/pg/policy.py]: yeah, this is copied from A3C so I should fix it there too

File: python/ray/rllib/pg/pg_evaluator.py
Comment (alvkao58) [reply python/ray/rllib/pg/pg_evaluator.py]: I think we still need self.config for the discount factor.

Comment (richardliaw) [reply python/ray/rllib/pg/pg_evaluator.py]: ah this comment was in reference for `self.registry` but has since been removed.

**Comment by richardliaw** [comment]
ok I think the last todos are 
1. add jenkins tests to make sure it works for some atari and classical control domain
2. make sure it hits 200 in CartPole-v0 and learns something in Acrobot-v0.

File: python/ray/rllib/pg/pg.py
Comment (richardliaw) [inline python/ray/rllib/pg/pg.py]: Let's call this `model` to keep consistency with A3C, PPO, etc

File: python/ray/rllib/pg/pg_evaluator.py
Comment (richardliaw) [inline python/ray/rllib/pg/pg_evaluator.py]: Let's remove this

**Review by richardliaw — CHANGES_REQUESTED** [review]
A few more minor nits. Note - this is also failing for Pong-v0, so it'd be nice to make sure that's fixed before this is merged.

For testing, can you add the following to `test/jenkins_tests/run_multi_node_tests.sh`:

```
docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
    python /ray/python/ray/rllib/train.py \
    --env CartPole-v0 \
    --run PG \
    --stop '{"training_iteration": 2}' \
    --config '{"batch_size": 500, "num_workers": 1}'

docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
    python /ray/python/ray/rllib/train.py \
    --env Pong-v0 \
    --run PG \
    --stop '{"training_iteration": 2}' \
    --config '{"batch_size": 500, "num_workers": 1}'

docker run --rm --shm-size=10G --memory=10G $DOCKER_SHA \
    python /ray/python/ray/rllib/train.py \
    --env FrozenLake-v0 \
    --run PG \
    --stop '{"training_iteration": 2}' \
    --config '{"batch_size": 500, "num_workers": 1}'
```

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
Looks good! I'll merge right after this last rerun passes.

**Pull request merged**
